File: src/test/java/blynx/thymeleaf/compositiondialect/testcomponents/Heading.java
package blynx.thymeleaf.compositiondialect.testcomponents;

import blynx.thymeleaf.compositiondialect.CompositionComponent;
import blynx.thymeleaf.compositiondialect.CompositionComponentContext;

/** A boxed {@code Integer}, not {@code int}: absent must be distinguishable from an explicit {@code 0}. */
public record Heading(Integer level, CompositionComponentContext context) implements CompositionComponent {

    public Heading {
        if (level == null) {
            level = context.variable("parentMagicHeadings") instanceof MagicHeadings parent
                    ? parent.level()
                    : 1;
        }
    }
}

File: dialect-lib/src/test/kotlin/blynx/thymeleaf/compositiondialect/testcomponents/Heading.kt
package blynx.thymeleaf.compositiondialect.testcomponents

import blynx.thymeleaf.compositiondialect.CompositionComponent
import blynx.thymeleaf.compositiondialect.CompositionComponentContext

class Heading(context: CompositionComponentContext) : CompositionComponent(context) {
    val level: Int = context.attributes["level"]?.toString()?.toIntOrNull()
        ?: (context.variable("parentMagicHeadings") as? MagicHeadings)?.level
        ?: 1
}
